README
GBROWSE README

The Generic Genome Browser (GBrowse) is a simple but highly
configurable web-based genome browser.  It is a component of the
Generic Model Organism Systems Database project (GMOD).

******* DOWNLOADING ******

The 2.X versions are now considered stable. You may download it from
CPAN at http://search.cpan.org/~lds/GBrowse/ or get the developer's
version via SVN:

Checkout read-only
git clone git://github.com/GMOD/GBrowse.git


The older 1.X version is still available for download here:

http://sourceforge.net/projects/gmod/files/Generic%20Genome%20Browser/GBrowse-1.70/Generic-Genome-Browser-1.70.tar.gz/download

or via SVN here:

svn co https://gmod.svn.sourceforge.net/svnroot/gmod/Generic-Genome-Browser/branches/stable Generic-Genome-Browser

**************************

SYNOPSIS:

  perl Build.PL
  ./Build distmeta
  ./Build test
  ./Build config
  ./Build demo    (optional)
  ./Build install (will need to become superuser)
  ./Build apache_conf

TO INSTALL:

Detailed install and configure instructions are available at
http://gmod.org/wiki/GBrowse_2.0_HOWTO and we suggest you follow
that. What follows here is a brief synopsis.

1) You will need a bunch of Perl prerequisites:

    Module                  Version
    ---------------------+----------
    perl                    5.008   
    Bio::Perl               1.007001
    CGI::Session            4.02    
    Bio::Graphics           2.09
    GD                      2.07    
    Digest::MD5             any
    File::Temp              any       
    IO::String              any       
    JSON                    any       
    LWP                     any       
    Storable                any
    Statistics::Descriptive any

    The Build script will tell you if you are missing any
    prerequisites. All of these are available on CPAN (www.cpan.org),
    or by using the CPAN shell ("perl -MCPAN -e shell").

- If you wish to take advantage of user accounts (which lets users share
  tracks with each other), you must have:

    Module                  Version
    ---------------------+----------
    DBI                     any
    Net::SMTP               any

  -and either-
    DBD::mysql              any
  -or-
    DBD::SQLite             any

- You should install BioPerl from CPAN (currently version 1.7.1) or, to
  take advantage of additional features and performance boosts, get a
  snapshot of the current developer version:

  http://github.com/bioperl/bioperl-live/tarball/master

2) Run the command:

     perl Build.PL

   This will create a script named "Build" in the current directory. If you
   have downloaded and unpacked, but not installed Bioperl, then you should
   specify its location using the -I option:

     perl -I /home/fred/build/bioperl-live Build.PL

   To install in a custom location, pass the --install_base (preferred) or
   --prefix argument (deprecated):

     perl Build.PL --install_base=/home/fred/gbrowse
     perl Build.PL --prefix=/home/fred/gbrowse

3) Run the command:

     ./Build test

   This will run a series of tests on the distribution. If one or more tests
   fail please inform the author.

4) Run the command:

    ./Build config

   This will ask you a series of questions about where Apache and its support
   files are located, and will allow you to choose where to install GBrowse's
   configuration files.

   If you change your mind and want to reconfigure GBrowse, simply run
   "./Build reconfig".

5) Run the command:

    ./Build demo

    This will launch configure an instance of Apache to run on a local port and 
    give you the URL for a GBrowse demo database. This allows you to try out the
    system before you install it.

6) Run the command:

    ./Build install

   You will probably need to be the superuser to do this. If you have "sudo" installed
   the invocation will be:
   
          % sudo ./Build install
          [sudo] password for jdoe: *****

   Otherwise

          % su
          Password: *****
          # ./Build install

7) Run the command:

       ./Build apache_config

   This will print out a fragment of an Apache configuration file that contains
   all the information needed to run GBrowse. Cut and paste this into your system's
   Apache configuration file and restart the server. Usually you will find GBrowse
   running at http://localhost/gbrowse2.

Support is available from the discussion groups and mailing lists at
www.gmod.org.

Lincoln Stein & the GMOD team
lincoln.stein@gmail.com
July 2010


--- COPYRIGHT AND LICENSING ---

Copyright (c) 2002 Cold Spring Harbor Laboratory and University of California, Berkeley
Copyright (c) 2010 Ontario Institute for Cancer Research

This package and its accompanying libraries is free software; you can
redistribute it and/or modify it under the terms of the GPL (either
version 1, or at your option, any later version) or the Artistic
License 2.0.  Refer to LICENSE for the full license text.

--- AUTHORS ---

This project has had many authors.  Some of the most prominent are:

Scott Cain <cain@cshl.edu>
Ben Faga <benfaga@gmail.com> 
Vsevolod (Simon) Ilyushchenko <simonf@cshl.edu>
Marc Logghe <Marc.Logghe@devgen.com>
Aaron Mackey <amackey@pcbi.upenn.edu>
Sheldon McKay <mckays@cshl.edu>
Nathan O. Siemers <Nathan.Siemers@bms.com>
Ian Davies <i.davies2@rogers.com>
Lincoln Stein <lincoln.stein@gmail.com>
Peter Ruzanov <peter.ruzanov@oicr.on.ca>
Chris Vandevelde <Christopher.Vandevelde@oicr.on.ca>
Tony deCatanzaro <tonydecat@gmail.com>
Sukhwinder Lall <sukhwinder.lall@oicr.on.ca>
Ahmed Einalem <aelnaiem@gmail.com>

Special thanks go to the authors of the translations, who include:

Franck Aniere <aniere@genoscope.cns.fr> (French)
Nansheng Chen <chen@cshl.edu> (Chinese)
Toshiaki Katayama <k@bioruby.org> (Japanese)
Marc Logghe <marcl@devgen.com> (Dutch)
Marco Mangone <mangone@cshl.edu> (Italian)
Linus Taejoon Kwon <linusben <at> bawi <dot> org> (Korean)
Marcela Tello-Ruiz <marcela@broad.mit.edu> (Spanish)
Gudmundur Thorisson <mummi@cshl.edu> (Icelandic)


Changes
2.56
 * SECURITY FIXES:
    - Prevent ability of an attacker to delete other users' accounts
    - Fix cookies to carry the httponly flag, reducing chance of cross-site
      scripting exploits.
    - Fix cross-site scripting vulnerability in generation of citation text.
 * Fix failure of feature summary message to appear on mousedown events.
 * If svg2pdf executable is in search path, use this in preference to inkscape
   (thanks to Nathan Weeks)
 * Fix left-align of citation text in "?" menu (thanks to Nathan Weeks)
 * Avoid accumulating many .inverted files in the bio_graphics_ff_UUID cache
   (thanks to Nathan Weeks).
 * Fix summary mode for features with no source tag (thanks to Nathan Weeks).
 * Don't spawn child processes if there are no uncached requests (thanks to
   Nathan Weeks yet again!)

2.55
 * Better detection and handling of render slaves that have gone now. If all slaves are
   down then throws an error (once) and then reverts to local rendering for a period of time.
 * Updated installer to give dire warning if installing into /usr/local/apache
   and there doesn't appear to be an apache instance there.
 * Support truetype fonts (need Bio::Graphics 2.33 or higher; set truetype=1 in GBrowse.conf, and
   optionally pass font names like "Helvetica-12:Italic" to the track "font" option). 
 * Clicking on the "Go" button when an annotation plugin is selected now turns on the corresponding track.
 * Clicking on the "Go" button when a filter plugin is selected opens up the configuration dialog.
 * Added Juan Tena's split_wig.pl script, which is useful for uploading WIG files from genomes with lots
   of scaffolds.
 * Fix performance problems when using in-memory databases composed of multiple GFF3 files. The fix is
   a hack that involves overwriting the Bio::DB::SeqFeature::Store::LoadHelper module!

2.54
 * Version 2.53 introduced a bad bug into track configuration such
   that semantic zooming no longer works for user-provided
   configuration. This has now been fixed.

2.53
 * Fix a variety of semantic zooming inconsistencies.
 * Code simplification in data source.

2.52
 * Avoid clipping the subtrack selection link on long track titles.
 * Require Bio::Graphics version 2.31 to avoid bug in wiggle xyplot generation.

2.51
 * Fix "can't call FETCH without object reference" errors during ./Build test.
 * Fix pod documentation for gbrowse_grow_cloud_vol.pl.
 * Remove gbrowse_attach_slaves.pl.
 * Volumes added to cloud instance have deleteOnTermination set to true.

2.50
 * Add gbrowse_import_ucsc_db.pl, which will create a fully functional starter database
   from a UCSC data source (e.g. 'hg19').  
 * Add VirtualBox and EC2 virtual machine support (see http://gmod.org/wiki/GBrowse2_VMs).
 * Prompt user before overwriting a customized config file.

2.49
  * Add new "metadata" option in track stanzas for making subtracks selectable via
    structured metadata; see http://gmod.org/wiki/Creating_and_Managing_Subtracks_with_GBrowse2
    for details.
  * Specify which database to use for search result karyotype display using following
    syntax:
           [builtin:karyotype]
           database = my_database
  * Fixed error reporting when uploading badly-formatted WIG files.
  * Do something sensible when user tries to upload a bigwig file.
  * Some support for UCSC-style uploads with "track type=bigWig bigDataUrl=http://something...".

2.48
  * Fix track-level genbank and fasta dumping.
  * Fix (mostly) the batch and sequence dumping plugins.

2.47
  * Fix bugs in demo mode.
  * Fix session locking bug; FastCGI performance subjectively improved.
  * Revert to using inline imagemaps when not running in an accelerated environment.
  * Require Bio::Graphics version 2.26, so that transparency is supported.

2.46
  * Support for transparent overlay tracks. These will appear as a user settable formatting option	
    for any track with subtracks. To make this default for a track set "bump = overlap" and consider
    setting "color_series = 1" in order for background colors to be assigned automatically;
  * When downloading vista tracks, default download format to "vista_wiggle".
  * Fix signal handling under FastCGI in order to kill runaway rendering processes
    when server restarted.

2.45
  * Restore h_feat & h_region functionality to gbrowse_img.
  * Fix gbrowse_img documentation to point to correct example URLs on WormBase.
  * Fix persistent grey screen when clicking on snapshot thumbnail.
  * Fix URL-escaped characters appearing in snapshot name.

2.44
  * Modest performance improvements when rendering multiple tracks.
  * Set absolute cap of 5000 features per track to avoid performance issues on NGS data.
    This cap can be changed by setting "maximum features" in the [GENERAL] section. Set
    to 0 to return to unlimited behavior.
  * Imagemap is now generated in such a way that time-consuming callbacks are only executed when
    mouse hovers or clicks on the feature. This noticeably speeds up track generation time. Drawback
    is that this increases the delay before balloon pops up on the hover operation when run in
    cgi-bin mode (still fast in FastCGI mode). Restore original behavior by setting 
    "inline imagemaps" (or "inline imagemap") to true in [TRACK DEFAULTS] 
    or individual track definitions.
  * Can now close tracks from main view even when corresponding track checkbox not showing
    in the track listing.
  * Default to flock locking instead of NFS locking; the former seems to be faster.
  * Enable shared session locking to improve performance.
  * Fix bad database pathnames in yeast configuration file examples.
  * Fix track finder plugin to avoid annoying extra refresh after leaving text entry field.
  * Fix broken sorting of subtracks.
  * Improve dynamic range of wiggle tracks.
  * mod_perl is no longer supported due to unfixable issues with zombie processes.

2.43

Build.PL
#!/usr/bin/perl

use strict;
use warnings;
use FindBin '$Bin';

use lib "$Bin/install_util";
use File::Copy;
use GBrowseInstall;

my $build = GBrowseInstall->new(
    module_name        => 'GBrowse',
    dist_version_from  => 'lib/Bio/Graphics/Browser2.pm',
    dist_author        => 'Lincoln Stein <lincoln.stein@gmail.com>',
    dist_abstract      => 'The GMOD Generic Genome Browser',
    license            => 'perl',

    requires     => {
	'perl'                => '5.008',
	'Bio::Root::Version'  => '1.007001',
	'Bio::Graphics'       => '2.34',
	'CGI::Session'        => '4.02',
	'Digest::MD5'         => 0,	
        'ExtUtils::CBuilder'  => 0,
	'File::Temp'          => 0,
	'GD'                  => '2.50',
	'Text::ParseWords'    => '3.27',
	'IO::String'          => 0,
	'JSON'                => 0,
	'LWP'                 => 0,
	'Statistics::Descriptive'  => 0,
	'Storable'            => 0,
        'Time::HiRes'         => 0,
        'Digest::SHA'         => 0,
        'Date::Parse'         => 0,
        'Term::ReadKey'       => 0,
	'HTTP::Daemon'        => 0,
	'parent'              => 0,
    },
    build_requires => {'Capture::Tiny'       => 0,},
    recommends  => {
	'Bio::Das'                 => 0,
	'Bio::DB::Sam'             => 1.36,
	'Bio::DB::BigFile'         => 1.00,
	'Crypt::SSLeay'            => 0,
	'DB_File::Lock'            => 0,
	'DBI'                      => 0,
	'DBD::mysql'               => 0,
	'DBD::Pg'                  => 0,
	'DBD::SQLite'              => 0,
	'Digest::SHA'              => 0,	
	'FCGI'                     => 0,
	'File::NFSLock'            => 0,
	'GD::SVG'                  => 0,
	'Math::BigInt'             => 0,
	'Net::OpenID::Consumer'    => 0,
	'Net::SMTP::SSL'           => 0,
	$] <= 5.008 ? ('Safe::World'  => 0)
                    : ()               ,
	'Template'                 => 2.20,
        'Term::ReadKey'            => 0,
	'VM::EC2'                  => 1.19,
	'Parse::Apache::ServerStatus' => 0,
    },
    script_files => [
	'bin/gbrowse_clean.pl',
	'bin/gbrowse_set_admin_passwd.pl',
	'bin/gbrowse_import_ucsc_db.pl',
	'bin/gbrowse_create_account.pl',
	'bin/gbrowse_change_passwd.pl',
	'bin/gbrowse_aws_balancer.pl',
	'bin/gbrowse_sync_aws_slave.pl',
	'bin/gbrowse_configure_slaves.pl',
	'bin/gbrowse_metadb_config.pl',
	'bin/gbrowse_grow_cloud_vol.pl',
	'bin/gbrowse_slave',
        'bin/gbrowse_syn_load_alignment_database.pl',
        'bin/gbrowse_syn_load_alignments_msa.pl',
	'bin/make_das_conf.pl',
	'bin/scan_gbrowse.pl',
	'bin/wiggle2gff3.pl',
	'bin/gtf2gff3.pl',
	'bin/bed2gff3.pl',
	'bin/ucsc_genes2gff.pl',
	'bin/load_genbank.pl',
    ],

    create_makefile_pl => 'passthrough',

    );

print STDERR <<END;
**** WARNING *****************************************************************************************
If you are upgrading from an earlier version of GBrowse and wish to use named user accounts, please 
set "user_accounts" to 1 in /etc/gbrowse2/GBrowse.conf and run bin/gbrowse_metadb_config.pl to 
create/update your user account database.

See /etc/gbrowse2/GBrowse.conf.new for suggested settings.
******************************************************************************************************
END

#print STDERR "Continue? [y/n] ";
#my $line = <>;
#exit 0 unless $line =~ /[yY]/;

if(-e '/proc/filesystems' 
   && `grep selinux /proc/filesystems`) {
    print STDERR <<'END';
**** WARNING *********************************************************
You appear to have SELinux installed on this computer. This can
interfere with GBrowse operation. Please read the file
README.fedora_and_selinux in order to perform additional configuration
operations that may be necessary to run on this computer.
**********************************************************************
END
}

if ($build->have_c_compiler()) {
    foreach ('CAlign.xs','CAlign.pm') {
	copy("./libalign/$_" => "./lib/Bio/Graphics/Browser/$_");
    }
    $build->include_dirs(["$Bin/libalign"]);
}

my $argc=0;
for my $arg (keys %{$build->private_props}) {
    my $value = $build->runtime_params($arg) or next;
    $build->config_data($arg=>$value);
    $argc++;
}
$build->config_data(OpenIDConsumerSecret=>int(1e15*rand()));
$build->config_done(1) if $argc > 0;

$build->add_build_element('conf');
$build->add_build_element('htdocs');
$build->add_build_element('cgibin');
$build->add_build_element('etc');
$build->add_build_element('database');

$build->create_build_script;
print STDERR <<END;
Now run:
  ./Build test
  ./Build demo          (optional)
  ./Build install       (as superuser/administrator)
        -or-
  ./Build install_slave (optional, for slave installations)

Other useful targets:
  ./Build register      (to register your copy of GBrowse)
  ./Build reconfig      (to reconfigure install locations)
  ./Build demostop      (to stop the demo)
END

exit 0;

1;
